File: tests/runtime/numbers/host.ts
// Flags: --instantiation

import * as helpers from "./helpers.js";
import { instantiate } from "./numbers.js";

function assertEq(x: any, y: any) {
  if (x !== y)
    throw new Error(`${x} != ${y}`);
}

function assert(x: boolean) {
  if (!x)
    throw new Error("assert failed");
}

async function run() {
  let scalar = 0;
  const wasm = await instantiate(helpers.loadWasm, {
    testwasi: helpers,
    imports: {
      roundtripU8(x) { return x; },
      roundtripS8(x) { return x; },
      roundtripU16(x) { return x; },
      roundtripS16(x) { return x; },
      roundtripU32(x) { return x; },
      roundtripS32(x) { return x; },
      roundtripU64(x) { return x; },
      roundtripS64(x) { return x; },
      roundtripFloat32(x) { return x; },
      roundtripFloat64(x) { return x; },
      roundtripChar(x) { return x; },
      setScalar(x) { scalar = x; },
      getScalar() { return scalar; },
    },
  });

  wasm.testImports();

  assertEq(wasm.exports.roundtripU8(1), 1);
  assertEq(wasm.exports.roundtripU8((1 << 8) - 1), (1 << 8) - 1);

  assertEq(wasm.exports.roundtripS8(1), 1);
  assertEq(wasm.exports.roundtripS8((1 << 7) - 1), (1 << 7) - 1);
  assertEq(wasm.exports.roundtripS8(-(1 << 7)), -(1 << 7));

  assertEq(wasm.exports.roundtripU16(1), 1);
  assertEq(wasm.exports.roundtripU16((1 << 16) - 1), (1 << 16) - 1);

  assertEq(wasm.exports.roundtripS16(1), 1);
  assertEq(wasm.exports.roundtripS16((1 << 15) - 1), (1 << 15) - 1);
  assertEq(wasm.exports.roundtripS16(-(1 << 15)), -(1 << 15));

  assertEq(wasm.exports.roundtripU32(1), 1);
  assertEq(wasm.exports.roundtripU32(~0 >>> 0), ~0 >>> 0);

  assertEq(wasm.exports.roundtripS32(1), 1);
  assertEq(wasm.exports.roundtripS32(((1 << 31) - 1) >>> 0), ((1 << 31) - 1) >>> 0);
  assertEq(wasm.exports.roundtripS32(1 << 31), 1 << 31);

  assertEq(wasm.exports.roundtripU64(1n), 1n);
  assertEq(wasm.exports.roundtripU64((1n << 64n) - 1n), (1n << 64n) - 1n);

  assertEq(wasm.exports.roundtripS64(1n), 1n);
  assertEq(wasm.exports.roundtripS64((1n << 63n) - 1n), (1n << 63n) - 1n);
  assertEq(wasm.exports.roundtripS64(-(1n << 63n)), -(1n << 63n));

  assertEq(wasm.exports.roundtripFloat32(1), 1);
  assertEq(wasm.exports.roundtripFloat32(Infinity), Infinity);
  assertEq(wasm.exports.roundtripFloat32(-Infinity), -Infinity);
  assert(Number.isNaN(wasm.exports.roundtripFloat32(NaN)));

  assertEq(wasm.exports.roundtripFloat64(1), 1);
  assertEq(wasm.exports.roundtripFloat64(Infinity), Infinity);
  assertEq(wasm.exports.roundtripFloat64(-Infinity), -Infinity);
  assert(Number.isNaN(wasm.exports.roundtripFloat64(NaN)));

  assertEq(wasm.exports.roundtripChar('a'), 'a');
  assertEq(wasm.exports.roundtripChar(' '), ' ');
  assertEq(wasm.exports.roundtripChar('🚩'), '🚩');

  wasm.exports.setScalar(2);
  assertEq(wasm.exports.getScalar(), 2);
  wasm.exports.setScalar(4);
  assertEq(wasm.exports.getScalar(), 4);
}

await run()

File: tests/runtime/numbers/host.py
import math
import wasmtime
from numbers import Numbers, NumbersImports
from helpers import TestWasi

class MyImports:
    def roundtrip_u8(self, a: int) -> int:
        return a

    def roundtrip_s8(self, a: int) -> int:
        return a

    def roundtrip_u16(self, a: int) -> int:
        return a

    def roundtrip_s16(self, a: int) -> int:
        return a

    def roundtrip_u32(self, a: int) -> int:
        return a

    def roundtrip_s32(self, a: int) -> int:
        return a

    def roundtrip_u64(self, a: int) -> int:
        return a

    def roundtrip_s64(self, a: int) -> int:
        return a

    def roundtrip_float32(self, a: float) -> float:
        return a

    def roundtrip_float64(self, a: float) -> float:
        return a

    def roundtrip_char(self, a: str) -> str:
        return a

    def set_scalar(self, a: int) -> None:
        self.scalar = a

    def get_scalar(self) -> int:
        return self.scalar

def run() -> None:
    store = wasmtime.Store()
    wasm = Numbers(store, NumbersImports(MyImports(), TestWasi()))

    wasm.test_imports(store)
    assert(wasm.roundtrip_u8(store, 1) == 1)
    assert(wasm.roundtrip_u8(store, (1 << 8) - 1) == (1 << 8) - 1)
    assert(wasm.roundtrip_u16(store, 1) == 1)
    assert(wasm.roundtrip_u16(store, (1 << 16) - 1) == (1 << 16) - 1)
    assert(wasm.roundtrip_u32(store, 1) == 1)
    assert(wasm.roundtrip_u32(store, (1 << 32) - 1) == (1 << 32) - 1)
    assert(wasm.roundtrip_u64(store, 1) == 1)
    assert(wasm.roundtrip_u64(store, (1 << 64) - 1) == (1 << 64) - 1)

    assert(wasm.roundtrip_s8(store, 1) == 1)
    assert(wasm.roundtrip_s8(store, (1 << (8 - 1) - 1)) == (1 << (8 - 1) - 1))
    assert(wasm.roundtrip_s8(store, -(1 << (8 - 1))) == -(1 << (8 - 1)))
    assert(wasm.roundtrip_s16(store, 1) == 1)
    assert(wasm.roundtrip_s16(store, (1 << (16 - 1) - 1)) == (1 << (16 - 1) - 1))
    assert(wasm.roundtrip_s16(store, -(1 << (16 - 1))) == -(1 << (16 - 1)))
    assert(wasm.roundtrip_s32(store, 1) == 1)
    assert(wasm.roundtrip_s32(store, (1 << (32 - 1) - 1)) == (1 << (32 - 1) - 1))
    assert(wasm.roundtrip_s32(store, -(1 << (32 - 1))) == -(1 << (32 - 1)))
    assert(wasm.roundtrip_s64(store, 1) == 1)
    assert(wasm.roundtrip_s64(store, (1 << (64 - 1) - 1)) == (1 << (64 - 1) - 1))
    assert(wasm.roundtrip_s64(store, -(1 << (64 - 1))) == -(1 << (64 - 1)))

    inf = float('inf')
    assert(wasm.roundtrip_float32(store, 1.0) == 1.0)
    assert(wasm.roundtrip_float32(store, inf) == inf)
    assert(wasm.roundtrip_float32(store, -inf) == -inf)
    assert(math.isnan(wasm.roundtrip_float32(store, float('nan'))))

    assert(wasm.roundtrip_float64(store, 1.0) == 1.0)
    assert(wasm.roundtrip_float64(store, inf) == inf)
    assert(wasm.roundtrip_float64(store, -inf) == -inf)
    assert(math.isnan(wasm.roundtrip_float64(store, float('nan'))))

    assert(wasm.roundtrip_char(store, 'a') == 'a')
    assert(wasm.roundtrip_char(store, ' ') == ' ')
    assert(wasm.roundtrip_char(store, '🚩') == '🚩')

    wasm.set_scalar(store, 2)
    assert(wasm.get_scalar(store) == 2)
    wasm.set_scalar(store, 4)
    assert(wasm.get_scalar(store) == 4)

if __name__ == '__main__':
    run()
